net/libfab/libfab.c

```diff
@@ -0,0 +1,2963 @@
+	int                 i;
+	int                 cnt;
+
+	memset(&buf, 0, sizeof(buf));
+	memset(&len, 0, sizeof(len));
+	if (cq != NULL) {
+		cnt = libfab_check_for_comp(cq, buf, len, data);
+		for (i = 0; i < cnt; i++) {
+			if (buf[i] != NULL) {
+				if (buf[i]->fb_length == 0)
+					buf[i]->fb_length = len[i];
+				buf[i]->fb_ev_ep = ep;
+				libfab_buf_done(buf[i], 0);
+			}
+			if (data[i] != 0)
+				libfab_buf_done((struct m0_fab__buf*)data[i],
+						0);
+		}
+	}
+}
+
+/**
+ * Check for completion events on the completion queue for the transmit endpoint
+ */
+static void libfab_txep_comp_read(struct fid_cq *cq)
+{
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__buf       *buf[FAB_MAX_COMP_READ];
+	int                       i;
+	int                       cnt;
+
+	memset(&buf, 0, sizeof(buf));
+	cnt = libfab_check_for_comp(cq, buf, NULL, NULL);
+	for (i = 0; i < cnt; i++) {
+		if (buf[i] != NULL) {
+			aep = libfab_aep_get(buf[i]->fb_txctx);
+			if (M0_IN(buf[i]->fb_nb->nb_qtype,
+				  (M0_NET_QT_ACTIVE_BULK_RECV,
+				   M0_NET_QT_ACTIVE_BULK_SEND)))
+				--aep->aep_bulk_cnt;
+			buf[i]->fb_wr_comp_cnt++;
+			if (buf[i]->fb_wr_comp_cnt >= buf[i]->fb_wr_cnt) {
+				libfab_target_notify(buf[i], aep);
+				libfab_buf_done(buf[i], 0);
+			}
+		}
+	}
+}
+
+/**
+ * Used to poll for connection events, completion events and process the queued
+ * bulk buffer operations.
+ */
+static void libfab_poller(struct m0_fab__tm *tm)
+{
+	struct m0_fab__ev_ctx    *ctx;
+	struct m0_fab__ep        *xep;
+	struct m0_fab__active_ep *aep;
+	struct fid_cq            *cq;
+	struct epoll_event        ev;
+	int                       ev_cnt;
+
+	/* Wait for the transfer machine to get started */
+	while (tm->ftm_state != FAB_TM_STARTED)
+		/* No-op */;
+	
+	while (tm->ftm_state != FAB_TM_SHUTDOWN) {
+		sched_yield();
+		ev_cnt = epoll_wait(tm->ftm_epfd, &ev, 1, FAB_WAIT_FD_TMOUT);
+
+		while (1) {
+			m0_mutex_lock(&tm->ftm_endlock);
+			if (tm->ftm_state == FAB_TM_SHUTDOWN)
+				break;
+			
+			if (libfab_tm_trylock(tm) != 0) {
+				m0_mutex_unlock(&tm->ftm_endlock);
+				/* the tm can not be locked now. 
+				   We will try it again */
+			} else {
+				/* Got tm lock. 
+				   Let's continue processing events. */
+				break;
+			}
+		}
+		
+		m0_mutex_unlock(&tm->ftm_endlock);
+		
+		if (tm->ftm_state == FAB_TM_SHUTDOWN)
+			break;
+		
+		M0_ASSERT(libfab_tm_is_locked(tm) && libfab_tm_invariant(tm));
+
+		if (ev_cnt > 0) {
+			ctx = ev.data.ptr;
+			if (ctx->evctx_type == FAB_COMMON_Q_EVENT) {
+				/* Check the common queue of the
+				   transfer machine for events */
+				libfab_handle_connect_request_events(tm);
+				libfab_txep_comp_read(tm->ftm_tx_cq);
+			} else {
+				/* Check the private queue of the
+				   endpoint for events */
+				xep = ctx->evctx_ep;
+				aep = libfab_aep_get(xep);
+				libfab_txep_event_check(xep, aep, tm);
+				cq = aep->aep_rx_res.frr_cq;
+				libfab_rxep_comp_read(cq, xep);
+			}
+		}
+
+		libfab_bulk_buf_process(tm);
```

Comment:
We should make sure to call this function _libfab_txep_comp_read()_ before every call to libfab_bulk_buf_process. This will free up the queue space and will make function libfab_bulk_buf_process to do more work